File: web/packages/engine/src/document.ts
import { compareStamp, STAMP_MIN, type Stamp } from "./hlc.js";
import { compareId, idKey, type ObjectId } from "./ids.js";
import { opObject, opStamp, type Op } from "./op.js";
import { valueAsString, type ObjectKind, type Value } from "./value.js";

export interface Register<T> {
  value: T;
  at: Stamp;
}

function merge<T>(reg: Register<T>, value: T, at: Stamp): boolean {
  if (compareStamp(at, reg.at) > 0) {
    reg.value = value;
    reg.at = at;
    return true;
  }
  return false;
}

export class ObjectState {
  /** `null` while the object is a ghost (ops seen, create not yet seen). */
  kind: ObjectKind | null = null;
  deleted: Register<boolean> = { value: false, at: STAMP_MIN };
  props = new Map<string, Register<Value>>();

  created(): boolean {
    return this.kind !== null;
  }

  visible(): boolean {
    return this.created() && !this.deleted.value;
  }

  get(key: string): Value | undefined {
    return this.props.get(key)?.value;
  }

  /** @internal */
  mergeProp(key: string, value: Value, at: Stamp): boolean {
    const reg = this.props.get(key);
    if (reg) return merge(reg, value, at);
    this.props.set(key, { value, at });
    return true;
  }
}

/**
 * Replicated document. `apply` is commutative, associative and idempotent;
 * see the Rust `Document` for the authoritative semantics.
 */
export class Document {
  private readonly objects = new Map<string, { id: ObjectId; state: ObjectState }>();

  apply(op: Op): boolean {
    const at = opStamp(op);
    const object = opObject(op);
    const key = idKey(object);
    let entry = this.objects.get(key);
    if (!entry) {
      entry = { id: object, state: new ObjectState() };
      this.objects.set(key, entry);
    }
    const s = entry.state;
    const k = op.kind;
    switch (k.op) {
      case "create": {
        let changed = false;
        if (s.kind === null) {
          s.kind = k.kind;
          changed = true;
        }
        for (const [pk, pv] of k.props) changed = s.mergeProp(pk, pv, at) || changed;
        return changed;
      }
      case "set_props": {
        let changed = false;
        for (const [pk, pv] of k.entries) changed = s.mergeProp(pk, pv, at) || changed;
        return changed;
      }
      case "delete":
        return merge(s.deleted, true, at);
      case "restore":
        return merge(s.deleted, false, at);
    }
  }

  get(id: ObjectId): ObjectState | undefined {
    return this.objects.get(idKey(id))?.state;
  }

  get size(): number {
    return this.objects.size;
  }

  /** All objects (ghosts and tombstones included) in canonical id order. */
  entries(): Array<[ObjectId, ObjectState]> {
    const out: Array<[ObjectId, ObjectState]> = [];
    for (const { id, state } of this.objects.values()) out.push([id, state]);
    out.sort((a, b) => compareId(a[0], b[0]));
    return out;
  }

  /** Visible objects in render order: by `z` (missing first), then id. */
  renderOrder(): Array<[ObjectId, ObjectState]> {
    const out = this.entries().filter(([, s]) => s.visible());
    out.sort(([ida, a], [idb, b]) => {
      const za = valueAsString(a.get("z"));
      const zb = valueAsString(b.get("z"));
      if (za === undefined && zb !== undefined) return -1;
      if (za !== undefined && zb === undefined) return 1;
      if (za !== undefined && zb !== undefined && za !== zb) return za < zb ? -1 : 1;
      return compareId(ida, idb);
    });
    return out;
  }

  /** @internal used by the snapshot decoder */
  insertRaw(id: ObjectId, state: ObjectState): void {
    this.objects.set(idKey(id), { id, state });
  }

  clone(): Document {
    const d = new Document();
    for (const [id, s] of this.entries()) {
      const c = new ObjectState();
      c.kind = s.kind;
      c.deleted = { ...s.deleted };
      for (const [k, r] of s.props) c.props.set(k, { ...r });
      d.insertRaw(id, c);
    }
    return d;
  }
}

File: crates/converge-core/src/document.rs
use std::collections::BTreeMap;

use crate::hlc::Stamp;
use crate::ids::ObjectId;
use crate::op::{Op, OpKind};
use crate::value::{ObjectKind, Value};

/// A last-writer-wins register.
#[derive(Clone, PartialEq, Debug)]
pub struct Register<T> {
    pub value: T,
    pub at: Stamp,
}

impl<T> Register<T> {
    pub fn new(value: T, at: Stamp) -> Self {
        Register { value, at }
    }

    /// Merge a write. Returns `true` if the register changed. Equal stamps
    /// only arise from the same op delivered twice, so they are a no-op.
    pub fn merge(&mut self, value: T, at: Stamp) -> bool {
        if at > self.at {
            self.value = value;
            self.at = at;
            true
        } else {
            false
        }
    }
}

#[derive(Clone, PartialEq, Debug)]
pub struct ObjectState {
    /// `None` while the object is a ghost (ops seen, `Create` not yet seen).
    pub kind: Option<ObjectKind>,
    pub deleted: Register<bool>,
    pub props: BTreeMap<String, Register<Value>>,
}

impl ObjectState {
    fn ghost() -> Self {
        ObjectState {
            kind: None,
            deleted: Register::new(false, Stamp::MIN),
            props: BTreeMap::new(),
        }
    }

    pub fn created(&self) -> bool {
        self.kind.is_some()
    }

    /// Rendered ⇔ created and not tombstoned.
    pub fn visible(&self) -> bool {
        self.created() && !self.deleted.value
    }

    pub fn get(&self, key: &str) -> Option<&Value> {
        self.props.get(key).map(|r| &r.value)
    }

    fn merge_prop(&mut self, key: &str, value: &Value, at: Stamp) -> bool {
        match self.props.get_mut(key) {
            Some(reg) => reg.merge(value.clone(), at),
            None => {
                self.props
                    .insert(key.to_string(), Register::new(value.clone(), at));
                true
            }
        }
    }
}

/// The replicated document state. `apply` is commutative, associative and
/// idempotent over ops, so every replica that has seen the same *set* of ops
/// holds byte-identical state (see [`crate::codec::canonical_bytes`]).
#[derive(Clone, PartialEq, Debug, Default)]
pub struct Document {
    objects: BTreeMap<ObjectId, ObjectState>,
}

impl Document {
    pub fn new() -> Self {
        Self::default()
    }

    /// Apply one op. Returns `true` if any state changed (false for
    /// duplicates and stale writes).
    pub fn apply(&mut self, op: &Op) -> bool {
        let at = op.stamp();
        let object = op.object();
        let entry = self
            .objects
            .entry(object)
            .or_insert_with(ObjectState::ghost);
        match &op.kind {
            OpKind::Create { kind, props } => {
                let mut changed = false;
                if entry.kind.is_none() {
                    entry.kind = Some(*kind);
                    changed = true;
                }
                for (k, v) in props {
                    changed |= entry.merge_prop(k, v, at);
                }
                changed
            }
            OpKind::SetProps { entries, .. } => {
                let mut changed = false;
                for (k, v) in entries {
                    changed |= entry.merge_prop(k, v, at);
                }
                changed
            }
            OpKind::Delete { .. } => entry.deleted.merge(true, at),
            OpKind::Restore { .. } => entry.deleted.merge(false, at),
        }
    }

    pub fn get(&self, id: &ObjectId) -> Option<&ObjectState> {
        self.objects.get(id)
    }

    /// All objects including ghosts and tombstones, in canonical order.
    pub fn objects(&self) -> impl Iterator<Item = (&ObjectId, &ObjectState)> {
        self.objects.iter()
    }

    pub fn len(&self) -> usize {
        self.objects.len()
    }

    pub fn is_empty(&self) -> bool {
        self.objects.is_empty()
    }

    /// Visible objects in render order: by `z` (fractional index, missing
    /// sorts first) then by id.
    pub fn render_order(&self) -> Vec<(&ObjectId, &ObjectState)> {
        let mut v: Vec<_> = self.objects.iter().filter(|(_, o)| o.visible()).collect();
        v.sort_by(|(ida, a), (idb, b)| {
            let za = a.get("z").and_then(Value::as_str);
            let zb = b.get("z").and_then(Value::as_str);
            za.cmp(&zb).then(ida.cmp(idb))
        });
        v
    }

    pub(crate) fn insert_raw(&mut self, id: ObjectId, state: ObjectState) {
        self.objects.insert(id, state);
    }
}
